dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+#define _VERBOUT_                                                              \
+  if (verbose)                                                                 \
+  std::cout
+#endif
+
+#ifndef _ERROUT_
+#define _ERROUT_ std::cerr
+#endif
+
+#ifndef MAX_SCENE_NODES
+#define MAX_SCENE_NODES 1000000
+#endif
+
+typedef struct {
+  int baseColorImage;
+  GLfloat metallicFactor;
+  GLfloat roughnessFactor;
+} SceneMaterial;
+
+SceneMaterial newSceneMaterial = {-1, 0.0f, 0.5f};
+
+typedef struct {
+  fgl::Mesh mesh;
+  int material;
+  int mode;
+} ScenePrim;
+
+typedef std::vector<ScenePrim> SceneMesh;
+
+typedef struct {
+  int mesh;
+  float3 translation;
+  float4 rotation;
+  float3 scale;
+  std::vector<int> children;
+} SceneNode;
+
+typedef struct {
+  int node;
+  int type;
+  int interpolation;
+  std::vector<float3> translations;
+  std::vector<float4> rotations;
+  std::vector<float3> scales;
+  std::vector<float> weights;
+} SceneAnimationData;
+
+typedef struct {
+  std::string name;
+  std::vector<SceneAnimationData> channels;
+} SceneAnimation;
+
+typedef struct {
+  std::vector<SceneMaterial> materials;
+  std::vector<SceneMesh> meshes;
+  std::vector<SceneNode> nodes;
+  std::vector<SceneAnimation> animations;
+  std::vector<int> roots;
+} Scene;
+
+Scene newScene = {{}, {}, {}, {}, {}};
+
+typedef struct {
+  std::string path;
+  fgl::Texture texture;
+} Image;
+
+std::vector<Image> images;
+
+int LoadImage(std::string gltfPath, std::string imgName, bool filter = true) {
+  // loads an image relative to the specified gltf path
+  // never loads the same image twice in the same process
+
+  // set the file path to the same folder as the glTF file
+  std::string imgFilePath =
+      gltfPath.substr(0, gltfPath.find_last_of('/') + 1) + imgName;
+  // try to find the image among already loaded images
+  for (size_t i = 0; i < images.size(); i++)
+    if (images[i].path == imgFilePath)
+      return i;
+  // no image found, so load the named file
+  int imgWidth, imgHeight, imgChannels;
+  std::string fileData;
+  if (LoadFile(imgFilePath.c_str(), &fileData) < 0) {
+    _ERROUT_ << "Failed to open " << imgFilePath << std::endl << std::endl;
+    return -1;
+  }
+  unsigned char *imgData =
+      stbi_load_from_memory((stbi_uc*)fileData.data(), fileData.size(), &imgWidth, &imgHeight, &imgChannels, 0);
+  fgl::Texture imgTexture =
+      fgl::LoadTexture(imgData, imgWidth, imgHeight, imgChannels, filter, filter);
+  stbi_image_free(imgData);
+  images.push_back({imgFilePath, imgTexture});
+  return images.size() - 1;
+}
+
+Scene LoadScene(std::string gltfPath, std::string schemaPath = "gltf.fbs") {
+  _VERBOUT_ << "GLTF SCENE" << std::endl;
+  try {
+    std::string gltfText;
+    if (LoadFile(gltfPath.c_str(), &gltfText) < 0) {
+      _ERROUT_ << "Failed to open " << gltfPath << std::endl << std::endl;
+      return newScene;
+    }
+
+    std::string schemaText;
+    if (LoadFile(schemaPath.c_str(), &schemaText) < 0) {
```

Comment:
by default, code should load the binary version, that does not require the schema or json parsing. You can then have a utility function that loads it via json + schema. In fact, this utility already exists, and is `flatbuffers_json_to_binary` in file.cpp. You can also leave the file loading to the user (see `read_file` in file.cpp), that way this function that just receive the FlatBuffer binary as a `string_view` (taken from the input LString::strv()`). That way they can choose if they want to load binary directly or go via json.